optuna/storages/_rdb/storage.py

```diff
@@ -1166,7 +1166,8 @@ def fail_stale_trials(self) -> List[int]:
         stale_trial_ids = self._get_stale_trial_ids()
 
         for trial_id in stale_trial_ids:
-            self.set_trial_state(trial_id, TrialState.FAIL)
+            if not self.set_trial_state(trial_id, TrialState.FAIL):
+                stale_trial_ids.remove(trial_id)
```

Comment:
Let's avoid changing the length of a list we're iterating over.

```python
lst = list(range(10))

for item in lst:
    # This changes the length of `lst` and should be avoided.
    lst.remove(item)

print(lst)  # [1, 3, 5, 7, 9], probably not what one intended.
```

We can construct a new list to return instead (or construct a copy in the beginning. Performance difference is probably negligible here).